File: Q1_Java/src/main/java/Product/Food/CannedFood.java
package Product.Food; // Package for Food products

import Product.Supplier;

// CannedFood class, extends Food
public class CannedFood extends Food {
    private String contents; // Type of food inside the can (e.g., "Tomato Soup", "Baked Beans")

    // Constructor for CannedFood
    public CannedFood(String id, String name, Supplier supplier, int packagingUnits, double price, String expirationDate, boolean isOrganic, String contents) {
        super(id, name, supplier, packagingUnits, price, expirationDate, isOrganic); // Calling Food's constructor
        this.contents = contents; // Setting the contents of the can
    }

    // Overriding calculateDiscount method
    @Override
    public double calculateDiscount(int quantity, int month) {
        double discountRate = 0.12; // Base discount of 12%
        if (month == 11) { // Additional discount in November
            discountRate += 0.05; // 5% more
        }
        return quantity * discountRate; // Calculating total discount
    }

    // Overriding displayProductInfo method
    @Override
    public String displayProductInfo() {
        // Returns string representation of CannedFood
        return "CannedFood{" +
               "ID='" + getId() + '\'' +
               ", Name='" + getName() + '\'' +
               ", Contents='" + contents + '\'' +
               ", Expiration Date='" + getExpirationDate() + '\'' +
               ", Is Organic=" + (isOrganic() ? "Yes" : "No") +
               ", Supplier='" + getSupplier().getName() + '\'' +
               ", Packaging Units=" + getPackagingUnits() +
               '}';
    }

    // Getter and setter for contents
    public String getContents() { return contents; }
    public void setContents(String contents) { this.contents = contents; }
}

File: Q3_Cpp/src/Product/Food/CannedFood.cpp
#include "../../../include/Product/Food/CannedFood.h"
#include <sstream>

// Constructor for CannedFood
CannedFood::CannedFood(std::string id, std::string name, Supplier supplier, int packagingUnits, double price, std::string expirationDate, bool isOrganic, std::string contents)
    : Food(id, name, supplier, packagingUnits, price, expirationDate, isOrganic), contents(contents) {
}

// Overriding calculateDiscount method from Product class
double CannedFood::calculateDiscount(int quantity, int month) {
    double discountRate = 0.12; // Base discount of 12%
    if (month == 11) { // Additional discount in November
        discountRate += 0.05; // 5% more
    }
    return quantity * price * discountRate; // Calculating total discount
}

// Overriding displayProductInfo method from Product class
std::string CannedFood::displayProductInfo() {
    // Returns string representation of CannedFood
    std::stringstream ss;
    ss << "CannedFood{" <<
       "ID='" << getId() << '\'' <<
       ", Name='" << getName() << '\'' <<
       ", Contents='" << contents << '\'' <<
       ", Expiration Date='" << getExpirationDate() << '\'' <<
       ", Is Organic=" << (getIsOrganic() ? "Yes" : "No") <<
       ", Supplier='" << getSupplier().getName() << '\'' <<
       ", Packaging Units=" << getPackagingUnits() <<
       '}';
    return ss.str();
}

// Getter and setter for contents
std::string CannedFood::getContents() const {
    return contents;
}

void CannedFood::setContents(const std::string& contents) {
    this->contents = contents;
}
